File: src/main/java/faba/java/engine.java
package faba.java;

import java.util.*;

final class ELattice<T extends Enum<T>> {
    final T bot;
    final T top;

    ELattice(T bot, T top) {
        this.bot = bot;
        this.top = top;
    }

    final T join(T x, T y) {
        if (x == bot) return y;
        if (y == bot) return x;
        if (x == y) return x;
        return top;
    }

    final T meet(T x, T y) {
        if (x == top) return y;
        if (y == top) return x;
        if (x == y) return x;
        return bot;
    }
}

final class Component<Id> {
    final boolean touched;
    final Set<Id> ids;

    Component(boolean touched, Set<Id> ids) {
        this.touched = touched;
        this.ids = ids;
    }

    Component<Id> remove(Id id) {
        if (ids.contains(id)) {
            HashSet<Id> newIds = new HashSet<Id>(ids);
            newIds.remove(id);
            return new Component<Id>(touched, newIds);
        }
        else {
            return this;
        }
    }

    Component<Id> removeAndTouch(Id id) {
        if (ids.contains(id)) {
            HashSet<Id> newIds = new HashSet<Id>(ids);
            newIds.remove(id);
            return new Component<Id>(true, newIds);
        } else {
            return this;
        }
    }

    boolean isEmpty() {
        return ids.isEmpty();
    }

    boolean isEmptyAndTouched() {
        return ids.isEmpty() && touched;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Component component = (Component) o;

        if (touched != component.touched) return false;
        if (!ids.equals(component.ids)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = (touched ? 1 : 0);
        result = 31 * result + ids.hashCode();
        return result;
    }
}

interface Result<Id, T> {}
final class Final<Id, T> implements Result<Id, T> {
    final T value;
    Final(T value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return "Final{" +
                "value=" + value +
                '}';
    }
}

final class Solution<Id, Val> {
    final Id id;
    final Val value;

    Solution(Id id, Val value) {
        this.id = id;
        this.value = value;
    }
}

final class Pending<Id, T> implements Result<Id, T> {
    final T infinum;
    final Set<Component<Id>> delta;

    Pending(T infinum, Set<Component<Id>> delta) {
        this.infinum = infinum;
        this.delta = delta;
    }
}

final class Equation<Id, T> {
    final Id id;
    final Result<Id, T> rhs;

    Equation(Id id, Result<Id, T> rhs) {
        this.id = id;
        this.rhs = rhs;
    }
}

final class Solver<Id, Val extends Enum<Val>> {
    private final ELattice<Val> lattice;
    private final HashMap<Id, Set<Id>> dependencies = new HashMap<>();
    private final HashMap<Id, Pending<Id, Val>> pending = new HashMap<>();
    private final Queue<Solution<Id, Val>> moving = new LinkedList<>();
    private final HashMap<Id, Val> solved = new HashMap<>();

    Solver(ELattice<Val> lattice) {
        this.lattice = lattice;
    }

    Solver(ELattice<Val> lattice, List<Equation<Id, Val>> equations) {
        this.lattice = lattice;
        for (Equation<Id, Val> equation : equations) {
            addEquation(equation);
        }
    }

    void addEquation(Equation<Id, Val> equation) {
        if (equation.rhs instanceof Final) {
            Final<Id, Val> finalResult = (Final<Id, Val>) equation.rhs;
            moving.add(new Solution<>(equation.id, finalResult.value));
        }
        else if (equation.rhs instanceof Pending) {
            Pending<Id, Val> pendingResult = (Pending<Id, Val>) equation.rhs;
            if (pendingResult.infinum.equals(lattice.top)) {
                moving.add(new Solution<>(equation.id, lattice.top));
            }
            else {
                for (Component<Id> component : pendingResult.delta) {
                    for (Id trigger : component.ids) {
                        Set<Id> set = dependencies.get(trigger);
                        if (set == null) {
                            set = new HashSet<>();
                            dependencies.put(trigger, set);
                        }
                        set.add(equation.id);
                    }
                }
                pending.put(equation.id, pendingResult);
            }
        }
    }

    Map<Id, Val> solve() {
        Solution<Id, Val> sol;
        while ((sol = moving.poll()) != null) {
            solved.put(sol.id, sol.value);
            Set<Id> dIds = dependencies.remove(sol.id);
            if (dIds != null) {
                for (Id dId : dIds) {
                    Pending<Id, Val> pend = pending.remove(dId);
                    if (pend != null) {
                        Result<Id, Val> pend1 = substitute(pend, sol.id, sol.value);
                        if (pend1 instanceof Final) {
                            Final<Id, Val> fi = (Final<Id, Val>) pend1;
                            moving.add(new Solution<>(dId, fi.value));
                        }
                        else {
                            pending.put(dId, (Pending<Id, Val>) pend1);
                        }
                    }
                }
            }
        }
        pending.clear();
        return solved;
    }

    Result<Id, Val> substitute(Pending<Id, Val> pending, Id id, Val value) {
        if (value.equals(lattice.bot)) {
            HashSet<Component<Id>> delta = new HashSet<>();
            for (Component<Id> component : pending.delta) {
                if (!component.ids.contains(id)) {
                    delta.add(component);
                }
            }
            if (delta.isEmpty()) {
                return new Final<>(pending.infinum);
            }
            else {
                return new Pending<>(pending.infinum, delta);
            }
        }
        else if (value.equals(lattice.top)) {
            HashSet<Component<Id>> delta = new HashSet<>();
            for (Component<Id> component : pending.delta) {
                Component<Id> component1 = component.remove(id);
                if (!component1.isEmptyAndTouched()) {
                    if (component1.isEmpty()) {
                        return new Final<>(lattice.top);
                    } else {
                        delta.add(component1);
                    }
                }
            }
            if (delta.isEmpty()) {
                return new Final<>(pending.infinum);
            }
            else {
                return new Pending<>(pending.infinum, delta);
            }
        }
        else {
            Val infinum = lattice.join(pending.infinum, value);
            if (infinum == lattice.top) {
                return new Final<>(lattice.top);
            }
            HashSet<Component<Id>> delta = new HashSet<>();
            for (Component<Id> component : pending.delta) {
                Component<Id> component1 = component.removeAndTouch(id);
                if (!component1.isEmpty()) {
                    delta.add(component1);
                }
            }
            if (delta.isEmpty()) {
                return new Final<>(infinum);
            }
            else {
                return new Pending<>(infinum, delta);
            }
        }
    }
}

File: src/main/scala/engine.scala
package faba.engine

import scala.collection.mutable

object `package` {
  implicit class LatticeOps[Val](x: Val)(implicit l: Lattice[Val]) {
    @inline def &(y: Val): Val =
      l.meet(x, y)

    @inline def meet(y: Val): Val =
      l.meet(x, y)

    @inline def |(y: Val): Val =
      l.join(x, y)

    @inline def join(y: Val): Val =
      l.join(x, y)
  }

  // sum of products
  type SoP[K, V] = Set[Component[K, V]]

  implicit class ResultOps[Id, Val](r1: Result[Id, Val])(implicit l: Lattice[Val]) {
    val top: Val = l.top
    @inline def join(r2: Result[Id, Val]): Result[Id, Val] = (r1, r2) match {
      case (Final(`top`), _) =>
        Final(`top`)
      case (_, Final(`top`)) =>
        Final(`top`)
      case (Final(v1), Final(v2)) =>
        Final(v1 join v2)
      case (Final(v1), Pending(comps2)) =>
        Pending(comps2 + Component(v1, Set()))
      case (Pending(comps1), Final(v2)) =>
        Pending(comps1 + Component(v2, Set()))
      case (Pending(comps1), Pending(comps2)) =>
        Pending(comps1 union comps2)
    }
  }
}

// complete finite lattice
trait Lattice[T] extends PartialOrdering[T] {
  val top: T
  val bot: T

  // |,  union
  final def join(x: T, y: T): T =
    (x, y) match {
      case (`bot`, _) => y
      case (_, `bot`) => x
      case (`top`, _) => top
      case (_, `top`) => top
      case _ => if (equiv(x, y)) x else top
    }

  // &, intersection
  final def meet(x: T, y: T): T =
    (x, y) match {
      case (`top`, _) => y
      case (_, `top`) => x
      case (`bot`, _) => bot
      case (_, `bot`) => bot
      case _ => if (equiv(x, y)) x else bot
    }

  final override def equiv(x: T, y: T): Boolean =
    x == y

  final override def tryCompare(x: T, y: T): Option[Int] =
    (x, y) match {
      case (`bot`, `bot`) => Some(0)
      case (_, `bot`) => Some(1)
      case (`bot`, _) => Some(-1)
      case (`top`, `top`) => Some(0)
      case (_, `top`) => Some(-1)
      case (`top`, _) => Some(1)
      case (_, _) => if (equiv(x, y)) Some(0) else None
    }

  final override def lteq(x: T, y: T): Boolean =
    tryCompare(x, y).getOrElse(1) <= 0
}

case class ELattice[E](bot: E, top: E) extends Lattice[E]

case class Component[Id, V](v: V, ids: Set[Id])

sealed trait Result[+Id, Val]
case class Final[Val](value: Val) extends Result[Nothing, Val]
case class Pending[Id, Val](delta: SoP[Id, Val]) extends Result[Id, Val]

case class Equation[Id, Val](id: Id, rhs: Result[Id, Val])

trait StableAwareId[K] {
  val stable: Boolean
  def mkUnstable: K
  def mkStable: K
}

final class Solver[K <: StableAwareId[K], V](implicit lattice: Lattice[V]) {
  type Solution = (K, V)
  val top = lattice.top
  val bot = lattice.bot

  private val dependencies = mutable.HashMap[K, Set[K]]()
  private val pending = mutable.HashMap[K, Pending[K, V]]()
  private val moving = mutable.Queue[Solution]()
  private var solved = Map[K, V]()

  def this(equations: List[Equation[K, V]])(implicit lattice: Lattice[V]) {
    this()
    equations.foreach(addEquation)
  }

  def addEquation(equation: Equation[K, V]): Unit =
    equation.rhs match {
      case Final(value) =>
        moving enqueue (equation.id -> value)
      case p@Pending(comps) =>
        for (trigger <- comps.map(_.ids).flatten) {
          dependencies(trigger) = dependencies.getOrElse(trigger, Set()) + equation.id
        }
        pending(equation.id) = p
    }

  def solve(): Map[K, V] = {
    while (moving.nonEmpty) {
      val (ident, value) = moving.dequeue()
      solved = solved + (ident -> value)

      // intricate logic here (null -> ... inference is a bit strange for now - optimistic assumption)
      val toPropagate: List[(K, V)] =
        if (ident.stable)
          List((ident, value), (ident.mkUnstable, value))
        else
          List((ident.mkStable, value), (ident, top))

      for {
        (pId, pValue) <- toPropagate
        dIds <- dependencies.remove(pId)
        dId <- dIds
        pend <- pending.remove(dId)
      } substitute(pend, pId, pValue) match {
        case Final(v) => moving enqueue (dId -> v)
        case p@Pending(_) => pending(dId) = p
      }
    }

    for ((id, _) <- pending)
      solved = solved + (id -> top)
    pending.clear()
    solved
  }

  private def substitute(pending: Pending[K, V], id: K, value: V): Result[K, V] = {
    val sum = pending.delta.map { prod =>
      if (prod.ids(id)) Component(value & prod.v, prod.ids - id) else prod
    }
    // can calculate now?
    var acc = bot
    for (Component(v, prod) <- sum) {
      v match {
        case `bot` =>
        case `top` if prod.isEmpty =>
          return Final(`top`)
        case _ =>
          if (prod.isEmpty) acc |= v else return Pending(sum)
      }
    }
    Final(acc)
  }

}
